### DISCUSSION TITLE
get word under MouseCursor or at Position of a word under the TextCursor, to display HelpScreen on F1 ?

### DISCUSSION THREAD

**Discussion by paule32** [magiblot/tvision#116] [Q&A] [answered]
Hello,
is it possible to get the word under the TextCursor, to make a context help that is catch from the name of the word and open the HelpWindow, and display Help Content for (keywords) ?

**Comment by magiblot** [comment, marked as answer]
Yes. The methods and fields you need are in `TEditor`. For example, to get the word under the cursor using a custom `TMyEditor` class which inherits from `TEditor`:

```c++
static bool isWhiteSpace(char c)
{
    return strchr("\t \n\r\0", c) == nullptr;
}

std::string TMyEditor::getWordUnderCursor()
{
    std::string s;

    uint wordEnd = nextWord(curPtr);
    uint wordBegin = prevWord(wordEnd);

    for (uint i = wordBegin; i < wordEnd; ++i)
    {
        char c = bufChar(i);
        if (!isWhiteSpace(c))
            s.push_back(c);
    }

    return s;
}
```

Now, the word under the cursor may change whenever the cursor position changes or the editor's contents get modified. Therefore you will need to update the help context in these cases. One possible way is to do it in the `handleEvent` method of the `TMyEditor` class.

**Reply by paule32** [reply]
F1 funcction key is invoked with a StatusBarMenu, or a TMenuBar/Item.
When I am have the Focus of a TEdit window, how can I get the word, when I press F1 ?
is that possible ?

**Reply by magiblot** [reply]
Let's say you assigned the F1 key to the command `cmHelp` (like the TVDemo application).

You want to get the currently focused `TMyEditor` (an inheritor of `TEditor`) so that you can get the word under its cursor.

In order to do this you can define a custom broadcast command, e.g. `cmGetFocusedEditor`, which will be handled by `TMyEditor`.

In `TMyEditor`'s constructor you must add:
```c++
eventMask |= evBroadcast;
```

And in `TMyEditor::handleEvent` you must add:
```c++
if (event.what == evBroadcast && event.message.command == cmGetFocusedEditor && (state & sfFocused))
    clearEvent(ev);
```

Then, wherever you are handling the `cmHelp` command (I guess in the `getEvent` method of your application class) you may add:
```c++
// Only the currently focused editor (if any) will answer to this command.
TMyEditor *editor = (TMyEditor *) message(deskTop, evBroadcast, cmGetFocusedEditor, 0);
if (editor)
{
    std::string word = editor->getWordUnderCursor();
}
```
